pyrealm/pmodel/two_leaf_irradience.py

```diff
@@ -299,3 +308,24 @@ def beta_angle(latitude: NDArray, declination: NDArray, hour_angle: NDArray) ->
     ) * np.cos(hour_angle)
 
     return beta
+
+
+def beta_angle_from_lat_long_time(
+    latitudes: NDArray, longitudes: NDArray, date_time: NDArray, fast: bool = False
+) -> NDArray:
+    """Calculate beta angle from position and time.
+
+    Uses Pysolar core libraries for more accurate calculation of beta (solar elevation)
+    NB: each element in date_time numpy array should be in pydatetime format.
+    """
+    solar_elevations = np.zeros_like(latitudes, dtype=float)
+
+    if not fast:
+        get_beta = get_altitude
+    else:
+        get_beta = get_altitude_fast
+
+    for i, (lat, lon, date_time) in enumerate(zip(latitudes, longitudes, date_time)):
+        solar_elevations[i] = get_beta(lat, lon, date_time)
```

Comment:
I don't think we can do this - partly the speed but also it doesn't generalise to multiple dimensions

```python
In [4]: import numpy as np
In [5]: x = np.arange(1,10).reshape((3,3))
In [6]: for ix, (x1, x2) in enumerate(zip(x, x)):
   ...:     print(ix, x1, x2)
   ...: 
0 [1 2 3] [1 2 3]
1 [4 5 6] [4 5 6]
2 [7 8 9] [7 8 9]
```
If I read the `pysolar` code correctly, I think the `get_altitude_fast` is the same (or similar) to the previous beta calculation. I'd love to outsource these calculations to a more specialised and precise package but maybe not now?

https://github.com/pingswept/pysolar/blob/ed75beb5ccfe9eeb2d4a6a410f5f953214e8d19f/pysolar/solar.py#L134

Reply:
Ah, OK. I think I made the assumption that the array would be 1D.

I just took a look at the pysolar functionality as I thought you were prompting that earlier in the discussion! No matter though.

Reply:
It would be great to use something like this -  especially if it opens up better models of atmosphere etc - but I think the ability to handle arrays is a prerequisite unfortunately. With a couple of exceptions, the code in `pyrealm` is written to handle arbitrary dimensionality, although the most usual use cases are: 1D (time series, typically), 2D (XY, or maybe a set of sites through time), 3D (XYT).

Reply:
The "exceptions" are really just - first axis must be Time because there is a lagged process involved.